daffodil-sapi/src/test/scala/org/apache/daffodil/example/TestScalaAPI.scala

```diff
@@ -1127,44 +1129,101 @@ class TestScalaAPI {
+      TestInfosetEvent.endComplex("e1", "http://example.com"),
+      TestInfosetEvent.endDocument()
+    )
+
+    val c = Daffodil.compiler()
+
+    val schemaFile = getResource("/test/sapi/mySchema1.dfdl.xsd")
+    val pf = c.compileFile(schemaFile)
+    val dp = pf.onPath("/")
+
+    val file = getResource("/test/sapi/myData.dat")
+    val fis = new java.io.FileInputStream(file)
+    val dis = new InputSourceDataInputStream(fis)
+    val outputter = new TestInfosetOutputter()
+    val pr = dp.parse(dis, outputter)
+
+    assertFalse(pr.isError())
+    assertArrayEquals(
+      expectedEvents.asInstanceOf[Array[Object]],
+      outputter.events.toArray.asInstanceOf[Array[Object]]
+    )
+
+    val bos = new java.io.ByteArrayOutputStream()
+    val wbc = java.nio.channels.Channels.newChannel(bos)
+    val inputter = new TestInfosetInputter(expectedEvents: _*)
+
+    val ur = dp.unparse(inputter, wbc)
+    assertFalse(ur.isError)
+    assertEquals(expectedData, bos.toString())
+  }
+
+  @Test
+  def testScalaAPICDATA1(): Unit = {
+    val expected = "<![CDATA[NO_WHITESPACE_AT_ALL]]>"
+    val data = "NO_WHITESPACE_AT_ALL$"
+    val schemaType = "string"
+    doXMLOutputStyleTest(expected, data, schemaType)
+  }
+
+  @Test
+  def testScalaAPICDATA2(): Unit = {
+    val expected = "<![CDATA[   'some' stuff   here &#xE000; and ]]]]><![CDATA[> even]]>"
+    val data = "   'some' stuff   here &#xE000; and ]]> even$"
+    val schemaType = "string"
+    doXMLOutputStyleTest(expected, data, schemaType)
+  }
+
+  @Test
+  def testScalaAPICDATA3(): Unit = {
+    val expected = "6.892"
+    val data = "6.892"
+    val schemaType = "float"
+    doXMLOutputStyleTest(expected, data, schemaType)
+  }
+
+  @Test
+  def testScalaAPICDATA4(): Unit = {
+    val expected = "<![CDATA[this contains a CRLF\nline ending]]>"
+    val data = "this contains a CRLF\r\nline ending$"
+    val schemaType = "string"
+    doXMLOutputStyleTest(expected, data, schemaType)
+  }
+
+  def doXMLOutputStyleTest(expect: String, data: String, schemaType: String): Unit = {
+    val c = Daffodil.compiler()
+    val schemaFile = schemaType match {
+      case "string" => getResource("/test/sapi/mySchemaCDATAString.dfdl.xsd")
+      case "float" => getResource("/test/sapi/mySchemaCDATAFloat.dfdl.xsd")
     }
+    val pf = c.compileFile(schemaFile)
+    var dp = pf.onPath("/")
+
+    val is = new ByteArrayInputStream(data.getBytes(StandardCharsets.UTF_8))
+    val input = new InputSourceDataInputStream(is)
+    val bosDP = new ByteArrayOutputStream()
+    val outputter = new XMLTextInfosetOutputter(bosDP, true, XMLOutputStyle.PrettyPrintSafe)
+    val res = dp.parse(input, outputter)
+    val err = res.isError()
+
+    val infosetDPString = bosDP.toString()
+    val start = infosetDPString.indexOf(".com\">") + 6
+    val end = infosetDPString.indexOf("</tns")
+    val value = infosetDPString.substring(start, end)
+
+    assertFalse(err)
+    assertEquals(expect, value)
+  }
 }
```

Comment:
Should we add similar JAPI tests to ensure this all works and doesn't accidentally break by a future change?

Reply:
Yes, let's add JAPI tests too.